File: haskell/src/Solutions/Euler215.hs
module Solutions.Euler215 where
import Control.Monad
import Data.Array
import Data.Array.ST

compute :: Int -> Int -> Int
compute bound maxH =
    let walls = let walls' n = case compare n bound of
                        LT -> map (n + 2 :) (walls' (n + 2)) ++ map (n + 3 :) (walls' (n + 3))
                        EQ -> [[]]
                        GT -> []
                in map init $ walls' 0
        l = length walls -- for bound = 32 it is 3329
        wallsArray :: Array Int [Int]
        wallsArray = listArray (1, l) walls
        crack :: [Int] -> [Int] -> Bool
        crack [] _ = True
        crack _ [] = True
        crack (h1 : t1) (h2 : t2) = case compare h1 h2 of
            LT -> crack t1 (h2 : t2)
            EQ -> False
            GT -> crack (h1 : t1) t2
        correct :: Array Int [Int]
        correct = listArray (1, l) [filter (\y -> crack (wallsArray ! x) (wallsArray ! y)) [1..l] | x <- [1..l]]
        -- bricks (h, x) = number of valid walls of height h with (crack ! x) at the top
        bricks :: Array (Int, Int) Int
        bricks = runSTArray $ do
            arr <- newArray ((1, 1), (maxH, l)) 1
            forM_ [2..maxH] $ \h -> forM_ [1..l] $ \x ->
                sum <$> forM (correct ! x) (\y -> readArray arr (h - 1, y)) >>= writeArray arr (h, x)
            return arr
    in sum $ map (\x -> bricks ! (maxH, x)) [1..l]

euler215 :: IO String
euler215 = return $ show $ compute 32 10

File: rust/src/solutions/euler215.rs
use cached::proc_macro::cached;
use std::cmp::Ordering;

fn walls(bound: usize) -> Vec<Vec<usize>> {
    #[cached]
    fn walls_(bound: usize, n: usize) -> Vec<Vec<usize>> {
        match n.cmp(&bound) {
            Ordering::Less => {
                let mut _2 = walls_(bound, n + 2);
                for row in &mut _2 {
                    row.push(n + 2);
                }
                let mut _3 = walls_(bound, n + 3);
                for row in &mut _3 {
                    row.push(n + 3);
                }
                _2.append(&mut _3);
                _2
            }
            Ordering::Equal => vec![vec![]],
            Ordering::Greater => vec![],
        }
    }
    let mut ans = walls_(bound, 0);
    for row in &mut ans {
        row.reverse();
        row.pop();
    }
    ans
}

fn crack(mut wall_1: &[usize], mut wall_2: &[usize]) -> bool {
    while !wall_1.is_empty() && !wall_2.is_empty() {
        match wall_1[0].cmp(&wall_2[0]) {
            Ordering::Less => wall_1 = &wall_1[1..],
            Ordering::Equal => return false,
            Ordering::Greater => wall_2 = &wall_2[1..],
        }
    }
    true
}

fn compute(bound: usize, max_h: usize) -> usize {
    let walls = walls(bound);
    let correct: Vec<Vec<usize>> = (0..walls.len())
        .map(|row| {
            (0..walls.len())
                .filter(|&i| crack(&walls[row][..], &walls[i][..]))
                .collect()
        })
        .collect();
    let mut bricks: Vec<Vec<usize>> = vec![vec![0; walls.len()]; max_h];
    for x in 0..walls.len() {
        bricks[0][x] = 1;
    }
    for h in 1..max_h {
        for x in 0..walls.len() {
            for &i in &correct[x] {
                bricks[h][x] += bricks[h - 1][i];
            }
        }
    }
    bricks[max_h - 1].iter().sum()
}

pub fn euler215() -> String {
    compute(32, 10).to_string()
}
